tools/events/ion_event_equivalence.cpp

```diff
@@ -0,0 +1,620 @@
+            ION_NEXT_ACTUAL_INDEX;
+            continue;
+        }
+        ION_ACCUMULATE_ASSERTION(ion_compare_events(ION_EVENT_EQUIVALENCE_ARGS));
+        ION_NEXT_VALUE_INDICES;
+    }
+    ION_EXIT_ASSERTIONS;
+}
+
+BOOL ion_compare_streams(IonEventStream *stream_expected, IonEventStream *stream_actual,
+                         IonEventResult *ION_RESULT_ARG) {
+    ION_ENTER_ASSERTIONS;
+    size_t ION_INDEX_EXPECTED_ARG = 0;
+    size_t ION_INDEX_ACTUAL_ARG = 0;
+    ION_EVENT_COMPARISON_TYPE ION_COMPARISON_TYPE_ARG = COMPARISON_TYPE_BASIC;
+    ION_ACCUMULATE_ASSERTION(
+            ion_compare_substreams(ION_EVENT_EQUIVALENCE_ARGS, stream_expected->size(), stream_actual->size()));
+    ION_EXIT_ASSERTIONS;
+}
+
+typedef BOOL (*COMPARISON_FN)(ION_EVENT_EQUIVALENCE_PARAMS);
+
+BOOL ion_compare_sets_equivs(ION_EVENT_EQUIVALENCE_PARAMS) {
+    ION_ENTER_ASSERTIONS;
+    ION_ACCUMULATE_ASSERTION(ion_compare_events(ION_EVENT_EQUIVALENCE_ARGS));
+    ION_EXIT_ASSERTIONS;
+}
+
+BOOL ion_compare_sets_nonequivs(ION_EVENT_EQUIVALENCE_PARAMS) {
+    ION_ENTER_ASSERTIONS;
+    // The corresponding indices are assumed to be equivalent.
+    if (ION_INDEX_EXPECTED_ARG != ION_INDEX_ACTUAL_ARG) {
+        ION_PREPARE_COMPARISON;
+        // Since inequality is expected here, passing a NULL result here prevents the comparison report from being
+        // polluted. If the events are equal, a comparison result stating such will be added to the report.
+        ION_RESULT_ARG = NULL;
+        ION_EXPECT_FALSE(ion_compare_events(ION_EVENT_EQUIVALENCE_ARGS), "Equivalent values in a non-equivs set.");
+    }
+    ION_EXIT_ASSERTIONS;
+}
+
+/**
+ * Compares each element in the current container to every other element in the container. The given index refers
+ * to the starting index of the first element in the container.
+ */
+BOOL ion_compare_sets_standard(ION_EVENT_EQUIVALENCE_PARAMS) {
+    ION_ENTER_ASSERTIONS;
+    COMPARISON_FN comparison_fn = (ION_COMPARISON_TYPE_ARG == COMPARISON_TYPE_EQUIVS
+                                   || ION_COMPARISON_TYPE_ARG == COMPARISON_TYPE_EQUIVTIMELINE)
+                                  ? ion_compare_sets_equivs
+                                  : ion_compare_sets_nonequivs;
+    const size_t index_actual_initial = ION_INDEX_ACTUAL_ARG;
+    while (TRUE) {
+        if (ION_GET_ACTUAL->event_type == CONTAINER_END && ION_GET_ACTUAL->depth == 0) {
+            ION_NEXT_EXPECTED_VALUE_INDEX;
+            if (ION_GET_EXPECTED->event_type == CONTAINER_END && ION_GET_EXPECTED->depth == 0) {
+                break;
+            }
+            ION_INDEX_ACTUAL_ARG = index_actual_initial;
+        } else {
+            ION_ACCUMULATE_ASSERTION((*comparison_fn)(ION_EVENT_EQUIVALENCE_ARGS));
+            ION_NEXT_ACTUAL_VALUE_INDEX;
+        }
+    }
+    ION_EXIT_ASSERTIONS;
+}
+
+/**
+ * The 'embedded_documents' annotation denotes that the current container contains streams of Ion data embedded
```

Comment:
It's not unique to ion-c; it's defined by ion-tests. See the ion-tests [README](https://github.com/amzn/ion-tests). It is just used for testing, but I agree that there is still the potential for naming conflicts. When it was created, it should have been prefixed with `$ion_` like all other reserved symbols. That's why I've proposed migrating it to `$ion_embedded_streams`, but this will require modifying all of the ion-tests vectors that contain the current `embedded_documents` annotation, as well as all of the implementations that have implemented the embedded streams test semantics.

Regarding your comment about the utility of allowing embedded documents in tests, here's an example where it's useful. The symbol `$ion_1_0`, when encountered as an unannotated value at the top level, receives the semantics of the Ion Version Marker. When annotated, or not at the top level, it is treated differently. In order to compare the equivalence or non-equivalence of different usages of `$ion_1_0`, it needs to be enclosed in a sequence type. However, as soon as you do that, it can no longer be at the top level. Embedding Ion streams allows the embedded stream to be read with a separate reader, and a separate symbol table context, which restarts at depth 0. That makes the following `non-equivs` example vector possible:
```
(
  "$ion_1_0"
  "annotated::$ion_1_0"
)
```

This is also particularly useful for testing the equivalence or non-equivalence of different symbol tables over the same stream.